# Fake News Classifier

The objective of this project is to build a classifier that is able to distinguish between the Fake (0) and Real(1) headline news.

There are two datasets provided:
- `training_data.csv`: Dataset provided to train and test the model.
- `testing_data.csv`: Dataset to provide the answer of the exercise.

`dataset/testing_data.csv` contains all labels set as 2. The purpose of the exercise is to replace them by 0 (fake) or 1 (real) according to the model predictions.

## 1. Environment Setup

### 1.1. Environment

In [1]:
import os
os.environ['TF_ENABLE_ONEDNN_OPTS'] = '0'

### 1.2. Tools and Libraries

In [2]:
import pandas as pd
import re
import sklearn
import xgboost as xgb
import warnings
from transformers import pipeline

/mnt/data/Projects/miniconda/envs/nltk2/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
2025-06-19 14:48:38.140744: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:467] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1750337318.149236   18956 cuda_dnn.cc:8579] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1750337318.151712   18956 cuda_blas.cc:1407] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
W0000 00:00:1750337318.158377   18956 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more

In [3]:
# Custom functions
%load_ext autoreload
%autoreload 2
from helpers import *

### 1.3. Data Loading and Preprocessing

In [4]:
warnings.filterwarnings("ignore")
seed = 42

## 2. Data Loading and Preprocessing

### 2.1. Initial Data Inspection

In [5]:
train_df = pd.read_csv("./dataset/training_data.csv", sep="\t", header=None)
train_df.columns = ["label", "corpus"]
train_df

label                                             corpus
0          0  donald trump sends out embarrassing new year‚s...
1          0  drunk bragging trump staffer started russian c...
2          0  sheriff david clarke becomes an internet joke ...
3          0  trump is so obsessed he even has obama‚s name ...
4          0  pope francis just called out donald trump duri...
...      ...                                                ...
34147      1  tears in rain as thais gather for late king's ...
34148      1  pyongyang university needs non-u.s. teachers a...
34149      1  philippine president duterte to visit japan ah...
34150      1  japan's abe may have won election\tbut many do...
34151      1  demoralized and divided: inside catalonia's po...

[34152 rows x 2 columns]

In [6]:
# Split between X and y
X = train_df["corpus"]
y = train_df["label"]

In [7]:
def find_all(find, text):
    result = re.findall(find, text)
    return result

print("How many JavaScript scripts are contained in the corpus?", X.apply(lambda x: len(find_all(r'<[/]?script[\s>]{1}', x.lower()))).sum())
print("How many CSS scripts are contained in the corpus?", X.apply(lambda x: len(find_all(r'<[/]?style[\s>]{1}', x.lower()))).sum())
print("How many HTML tags are contained in the corpus?", X.apply(lambda x: len(find_all(r'<[/a-zA-Z]{1,}[\s>]{1}', x))).sum())

How many JavaScript scripts are contained in the corpus? 0
How many CSS scripts are contained in the corpus? 0
How many HTML tags are contained in the corpus? 0


**Analysis:**
- We don't need to clean HTML, Javascript or HTML code

###  2.2. Data Cleaning

In [8]:
def clean_serie(serie):
    # Remove numbers and special numbers
    serie = serie.replace(r"[^a-zA-Z]", " ", regex=True, inplace=False)
    # Remove single chars
    serie = serie.replace(r'^(\w{1}\s)|(\s\w{1}\s)|(\s\w{1})$', " ", regex=True, inplace=False)
    # Remove multiple empty spaces
    serie = serie.replace(r"\s{2,}", " ", regex=True, inplace=False)
    # Make all string lowercase
    serie = serie.str.lower()
    return serie

X = clean_serie(X)

checkline = 3
print("Original:", train_df["corpus"][checkline])
print("Cleaned: ", X[checkline])

Original: trump is so obsessed he even has obama‚s name coded into his website (images)
Cleaned:  trump is so obsessed he even has obama name coded into his website images 


**Analysis**:
- We decided to format the data by:
    - Removing special characters and numbers
    - Remove single chars
    - Remove multiple spaces
    - Transform all the text to lowercase

###  2.3. Training and Test Sets

In [9]:
# Split Data between train and test
X_train, X_test, y_train, y_test = sklearn.model_selection.train_test_split(X, y, test_size=.2, random_state=seed)

# Prepare CountVector
count_vector = sklearn.feature_extraction.text.CountVectorizer(ngram_range=(2,2))

# Transform fit and transform data
train_ds = count_vector.fit_transform(X_train)
test_ds = count_vector.transform(X_test)

## 3. Baseline Model

### 3.1. Baseline Model

In [10]:
# Set and fit Classifier Model
rfc = sklearn.ensemble.RandomForestClassifier(n_estimators=200, n_jobs=-1, criterion='entropy', random_state=seed)

### 3.2. Training and Evaluation

In [11]:
# Fit the model
rfc.fit(train_ds, y_train)

# Get Test predictions
predictions = rfc.predict(test_ds)

result = get_classifier_metrics(y_test, predictions , sub_name="_test")
display(pd.DataFrame.from_dict(result, orient='index').T)

print("\nMetrics Details:")
print(sklearn.metrics.classification_report(y_test, predictions))

accuracy_test  precision_test  recall_test   f1_test
0       0.814668        0.871804     0.722895  0.790397


Metrics Details:
              precision    recall  f1-score   support

           0       0.78      0.90      0.83      3529
           1       0.87      0.72      0.79      3302

    accuracy                           0.81      6831
   macro avg       0.82      0.81      0.81      6831
weighted avg       0.82      0.81      0.81      6831



## 4. Machine Learning Model Experimentation

### 4.1. Models and Vectorizers Combinations

We'll compare the next models with different classifiers to see how do they perform (on default settings) with our data.

We'll focus on Accuracy and Recall since the last one is the metric we want to improve in order to avoid False Negatives (we will penalize Fake News classified as Real News)

In [12]:
models = [sklearn.neighbors.KNeighborsClassifier(),
          sklearn.linear_model.LogisticRegression(random_state=seed),
          sklearn.tree.DecisionTreeClassifier(random_state=seed),
          sklearn.ensemble.RandomForestClassifier(random_state=seed, n_jobs=-1),
          sklearn.ensemble.AdaBoostClassifier(random_state=seed),
          xgb.XGBClassifier(random_state=seed),
          sklearn.naive_bayes.BernoulliNB()]

#### 4.1.1. Count Vectorizer

In [13]:
vectorizers = [sklearn.feature_extraction.text.CountVectorizer()]
results = compare_vectorizers(vectorizers, models, X_train, y_train, X_test, y_test)
display(results)

model  fit_time       vectorizer  accuracy_train  \
0    KNeighborsClassifier  0.000908  CountVectorizer        0.607298   
1      LogisticRegression  4.545018  CountVectorizer        0.981223   
2  DecisionTreeClassifier  2.364206  CountVectorizer        1.000000   
3  RandomForestClassifier  1.123251  CountVectorizer        1.000000   
4      AdaBoostClassifier  0.562853  CountVectorizer        0.773508   
5           XGBClassifier  1.214446  CountVectorizer        0.916841   
6             BernoulliNB  0.002997  CountVectorizer        0.952710   

   precision_train  recall_train  f1_train  accuracy_test  precision_test  \
0         0.955016      0.201461  0.332732       0.573854        0.896552   
1         0.976199      0.985389  0.980773       0.949349        0.939619   
2         1.000000      1.000000  1.000000       0.879373        0.881701   
3         1.000000      1.000000  1.000000       0.929732        0.938471   
4         0.694566      0.953080  0.803543       0.782316        0.701086   
5         0.880672      0.958804  0.918079       0.908066        0.874510   
6         0.939756      0.964528  0.951981       0.944810        0.933333   

   recall_test   f1_test  
0     0.133858  0.232938  
1     0.956693  0.948079  
2     0.866747  0.874160  
3     0.914597  0.926380  
4     0.958207  0.809725  
5     0.945488  0.908615  
6     0.953967  0.943538

**Analysis**:
- The 2 best behaving models for our data and CountVectorizer are:
    - LogisticRegression (accuracy: 0.944, recall: 0.956)
    - BernoulliNB        (accuracy: 0.937, recall: 0.944)

#### 4.1.2. TF-IDF Vectorizer

In [14]:
vectorizers = [sklearn.feature_extraction.text.TfidfVectorizer()] # max_features=1000, stop_words="english"
result = compare_vectorizers(vectorizers, models, X_train, y_train, X_test, y_test)

display(result)
results = pd.concat([results, result], ignore_index=True)

model   fit_time       vectorizer  accuracy_train  \
0    KNeighborsClassifier   0.000936  TfidfVectorizer        0.930383   
1      LogisticRegression   1.947409  TfidfVectorizer        0.961751   
2  DecisionTreeClassifier   2.257914  TfidfVectorizer        1.000000   
3  RandomForestClassifier   0.985260  TfidfVectorizer        1.000000   
4      AdaBoostClassifier   1.547843  TfidfVectorizer        0.791442   
5           XGBClassifier  12.197378  TfidfVectorizer        0.932470   
6             BernoulliNB   0.003238  TfidfVectorizer        0.952710   

   precision_train  recall_train  f1_train  accuracy_test  precision_test  \
0         0.919407      0.939072  0.929136       0.894305        0.875655   
1         0.952973      0.969122  0.960980       0.944810        0.931798   
2         1.000000      1.000000  1.000000       0.883326        0.872214   
3         1.000000      1.000000  1.000000       0.934417        0.924450   
4         0.715537      0.947582  0.815372       0.798565        0.719963   
5         0.903110      0.964528  0.932809       0.918021        0.886633   
6         0.939756      0.964528  0.951981       0.944810        0.933333   

   recall_test   f1_test  
0     0.910660  0.892815  
1     0.955784  0.943639  
2     0.888855  0.880456  
3     0.941248  0.932773  
4     0.954573  0.820833  
5     0.952150  0.918224  
6     0.953967  0.943538

**Analysis**:
- The 2 best performing model for TfidfVectorizer are:
    - LogisticRegression  (accuracy: 0.937, recall: 0.950)
    - BernoulliNB         (accuracy: 0.937, recall: 0.944)
- RandomForestClassifier is a good candidate to be considered since has results very close to BernoulliNB
- While these results are not far, CountVector seems to perform a bit better without changing configurations.

#### 4.1.3. Hashing Vectorizer

In [15]:
vectorizers = [sklearn.feature_extraction.text.TfidfVectorizer()]
result = compare_vectorizers(vectorizers, models, X_train, y_train, X_test, y_test)

display(result)
results = pd.concat([results, result], ignore_index=True)

model   fit_time       vectorizer  accuracy_train  \
0    KNeighborsClassifier   0.000849  TfidfVectorizer        0.930383   
1      LogisticRegression   1.781562  TfidfVectorizer        0.961751   
2  DecisionTreeClassifier   2.279669  TfidfVectorizer        1.000000   
3  RandomForestClassifier   0.986437  TfidfVectorizer        1.000000   
4      AdaBoostClassifier   1.555851  TfidfVectorizer        0.791442   
5           XGBClassifier  10.548115  TfidfVectorizer        0.932470   
6             BernoulliNB   0.002911  TfidfVectorizer        0.952710   

   precision_train  recall_train  f1_train  accuracy_test  precision_test  \
0         0.919407      0.939072  0.929136       0.894305        0.875655   
1         0.952973      0.969122  0.960980       0.944810        0.931798   
2         1.000000      1.000000  1.000000       0.883326        0.872214   
3         1.000000      1.000000  1.000000       0.934417        0.924450   
4         0.715537      0.947582  0.815372       0.798565        0.719963   
5         0.903110      0.964528  0.932809       0.918021        0.886633   
6         0.939756      0.964528  0.951981       0.944810        0.933333   

   recall_test   f1_test  
0     0.910660  0.892815  
1     0.955784  0.943639  
2     0.888855  0.880456  
3     0.941248  0.932773  
4     0.954573  0.820833  
5     0.952150  0.918224  
6     0.953967  0.943538

In [16]:
results.to_csv("./output/improve_pred_results.csv")
display(results)

model   fit_time       vectorizer  accuracy_train  \
0     KNeighborsClassifier   0.000908  CountVectorizer        0.607298   
1       LogisticRegression   4.545018  CountVectorizer        0.981223   
2   DecisionTreeClassifier   2.364206  CountVectorizer        1.000000   
3   RandomForestClassifier   1.123251  CountVectorizer        1.000000   
4       AdaBoostClassifier   0.562853  CountVectorizer        0.773508   
5            XGBClassifier   1.214446  CountVectorizer        0.916841   
6              BernoulliNB   0.002997  CountVectorizer        0.952710   
7     KNeighborsClassifier   0.000936  TfidfVectorizer        0.930383   
8       LogisticRegression   1.947409  TfidfVectorizer        0.961751   
9   DecisionTreeClassifier   2.257914  TfidfVectorizer        1.000000   
10  RandomForestClassifier   0.985260  TfidfVectorizer        1.000000   
11      AdaBoostClassifier   1.547843  TfidfVectorizer        0.791442   
12           XGBClassifier  12.197378  TfidfVectorizer        0.932470   
13             BernoulliNB   0.003238  TfidfVectorizer        0.952710   
14    KNeighborsClassifier   0.000849  TfidfVectorizer        0.930383   
15      LogisticRegression   1.781562  TfidfVectorizer        0.961751   
16  DecisionTreeClassifier   2.279669  TfidfVectorizer        1.000000   
17  RandomForestClassifier   0.986437  TfidfVectorizer        1.000000   
18      AdaBoostClassifier   1.555851  TfidfVectorizer        0.791442   
19           XGBClassifier  10.548115  TfidfVectorizer        0.932470   
20             BernoulliNB   0.002911  TfidfVectorizer        0.952710   

    precision_train  recall_train  f1_train  accuracy_test  precision_test  \
0          0.955016      0.201461  0.332732       0.573854        0.896552   
1          0.976199      0.985389  0.980773       0.949349        0.939619   
2          1.000000      1.000000  1.000000       0.879373        0.881701   
3          1.000000      1.000000  1.000000       0.929732        0.938471   
4          0.694566      0.953080  0.803543       0.782316        0.701086   
5          0.880672      0.958804  0.918079       0.908066        0.874510   
6          0.939756      0.964528  0.951981       0.944810        0.933333   
7          0.919407      0.939072  0.929136       0.894305        0.875655   
8          0.952973      0.969122  0.960980       0.944810        0.931798   
9          1.000000      1.000000  1.000000       0.883326        0.872214   
10         1.000000      1.000000  1.000000       0.934417        0.924450   
11         0.715537      0.947582  0.815372       0.798565        0.719963   
12         0.903110      0.964528  0.932809       0.918021        0.886633   
13         0.939756      0.964528  0.951981       0.944810        0.933333   
14         0.919407      0.939072  0.929136       0.894305        0.875655   
15         0.952973      0.969122  0.960980       0.944810        0.931798   
16         1.000000      1.000000  1.000000       0.883326        0.872214   
17         1.000000      1.000000  1.000000       0.934417        0.924450   
18         0.715537      0.947582  0.815372       0.798565        0.719963   
19         0.903110      0.964528  0.932809       0.918021        0.886633   
20         0.939756      0.964528  0.951981       0.944810        0.933333   

    recall_test   f1_test  
0      0.133858  0.232938  
1      0.956693  0.948079  
2      0.866747  0.874160  
3      0.914597  0.926380  
4      0.958207  0.809725  
5      0.945488  0.908615  
6      0.953967  0.943538  
7      0.910660  0.892815  
8      0.955784  0.943639  
9      0.888855  0.880456  
10     0.941248  0.932773  
11     0.954573  0.820833  
12     0.952150  0.918224  
13     0.953967  0.943538  
14     0.910660  0.892815  
15     0.955784  0.943639  
16     0.888855  0.880456  
17     0.941248  0.932773  
18     0.954573  0.820833  
19     0.952150  0.918224  
20     0.953967  0.943538

**Analysis**:
- The 2 best performing model for TfidfVectorizer are:
    - RandomForestClassifier (accuracy: 0.948, recall: 0.955): 
    - LogisticRegression (accuracy: 0.929, recall: 0.947):
- Hashing Vectorizer together with Random Forest Generator show the best results for accuracy but slightly less recall than Logistic Regressor with Count Vetorizer.

### 4.2 N-Gram Analysis for the Top-Performing Models

#### 4.2.1. LogisticRegression + CountVectorizer

In [17]:
models = [sklearn.linear_model.LogisticRegression(random_state=seed)]
ngrams = [(1,1), (1,2), (2,2), (1,3), (2,3), (3,3)]
vectorizers = [sklearn.feature_extraction.text.CountVectorizer(ngram_range=ngram) for ngram in ngrams]

results = compare_vectorizers(vectorizers, models, X_train, y_train, X_test, y_test, inc_params=["ngram_range"])
display(results)

model  fit_time       vectorizer ngram_range  accuracy_train  \
0  LogisticRegression  4.904252  CountVectorizer      (1, 1)        0.981223   
1  LogisticRegression  5.254913  CountVectorizer      (1, 2)        0.998243   
2  LogisticRegression  3.619610  CountVectorizer      (2, 2)        0.995352   
3  LogisticRegression  4.721386  CountVectorizer      (1, 3)        0.999378   
4  LogisticRegression  4.223055  CountVectorizer      (2, 3)        0.997950   
5  LogisticRegression  5.531452  CountVectorizer      (3, 3)        0.995205   

   precision_train  recall_train  f1_train  accuracy_test  precision_test  \
0         0.976199      0.985389  0.980773       0.949349        0.939619   
1         0.996845      0.999548  0.998195       0.952862        0.940828   
2         0.991626      0.998870  0.995235       0.901186        0.864962   
3         0.998721      1.000000  0.999360       0.950227        0.937389   
4         0.995800      1.000000  0.997896       0.897087        0.856516   
5         0.990230      1.000000  0.995091       0.769580        0.686932   

   recall_test   f1_test  
0     0.956693  0.948079  
1     0.963053  0.951811  
2     0.942762  0.902188  
3     0.961236  0.949163  
4     0.945488  0.898805  
5     0.961538  0.801363

**Analysis**:
- Best results for LogisticRegression and CountVectorizer is use unigrams and bigrams (accuracy: 0.950, recall: 0.961)

#### 4.2.2. RandomForestClassifier + HashingVectorizer

In [18]:
models = [sklearn.ensemble.RandomForestClassifier(random_state=seed, n_jobs=-1)]
ngrams = [(1,1), (1,2), (2,2), (1,3), (2,3), (3,3)]
vectorizers = [sklearn.feature_extraction.text.TfidfVectorizer(ngram_range=ngram) for ngram in ngrams]

result = compare_vectorizers(vectorizers, models, X_train, y_train, X_test, y_test, inc_params=["ngram_range"])

display(result)
results = pd.concat([results, result], ignore_index=True)

model   fit_time       vectorizer ngram_range  \
0  RandomForestClassifier   1.005639  TfidfVectorizer      (1, 1)   
1  RandomForestClassifier   4.583967  TfidfVectorizer      (1, 2)   
2  RandomForestClassifier   6.214324  TfidfVectorizer      (2, 2)   
3  RandomForestClassifier  10.037241  TfidfVectorizer      (1, 3)   
4  RandomForestClassifier  27.188731  TfidfVectorizer      (2, 3)   
5  RandomForestClassifier  15.372178  TfidfVectorizer      (3, 3)   

   accuracy_train  precision_train  recall_train  f1_train  accuracy_test  \
0        1.000000              1.0      1.000000  1.000000       0.934417   
1        1.000000              1.0      1.000000  1.000000       0.913043   
2        0.999817              1.0      0.999623  0.999812       0.805299   
3        1.000000              1.0      1.000000  1.000000       0.904699   
4        0.999780              1.0      0.999548  0.999774       0.798273   
5        0.998902              1.0      0.997741  0.998869       0.636364   

   precision_test  recall_test   f1_test  
0        0.924450     0.941248  0.932773  
1        0.889528     0.936402  0.912364  
2        0.856988     0.716838  0.780673  
3        0.885208     0.922471  0.903455  
4        0.854197     0.702604  0.771020  
5        0.904950     0.276802  0.423933

In [19]:
results.to_csv("./output/n-grams_results.csv")
display(results)

model   fit_time       vectorizer ngram_range  \
0       LogisticRegression   4.904252  CountVectorizer      (1, 1)   
1       LogisticRegression   5.254913  CountVectorizer      (1, 2)   
2       LogisticRegression   3.619610  CountVectorizer      (2, 2)   
3       LogisticRegression   4.721386  CountVectorizer      (1, 3)   
4       LogisticRegression   4.223055  CountVectorizer      (2, 3)   
5       LogisticRegression   5.531452  CountVectorizer      (3, 3)   
6   RandomForestClassifier   1.005639  TfidfVectorizer      (1, 1)   
7   RandomForestClassifier   4.583967  TfidfVectorizer      (1, 2)   
8   RandomForestClassifier   6.214324  TfidfVectorizer      (2, 2)   
9   RandomForestClassifier  10.037241  TfidfVectorizer      (1, 3)   
10  RandomForestClassifier  27.188731  TfidfVectorizer      (2, 3)   
11  RandomForestClassifier  15.372178  TfidfVectorizer      (3, 3)   

    accuracy_train  precision_train  recall_train  f1_train  accuracy_test  \
0         0.981223         0.976199      0.985389  0.980773       0.949349   
1         0.998243         0.996845      0.999548  0.998195       0.952862   
2         0.995352         0.991626      0.998870  0.995235       0.901186   
3         0.999378         0.998721      1.000000  0.999360       0.950227   
4         0.997950         0.995800      1.000000  0.997896       0.897087   
5         0.995205         0.990230      1.000000  0.995091       0.769580   
6         1.000000         1.000000      1.000000  1.000000       0.934417   
7         1.000000         1.000000      1.000000  1.000000       0.913043   
8         0.999817         1.000000      0.999623  0.999812       0.805299   
9         1.000000         1.000000      1.000000  1.000000       0.904699   
10        0.999780         1.000000      0.999548  0.999774       0.798273   
11        0.998902         1.000000      0.997741  0.998869       0.636364   

    precision_test  recall_test   f1_test  
0         0.939619     0.956693  0.948079  
1         0.940828     0.963053  0.951811  
2         0.864962     0.942762  0.902188  
3         0.937389     0.961236  0.949163  
4         0.856516     0.945488  0.898805  
5         0.686932     0.961538  0.801363  
6         0.924450     0.941248  0.932773  
7         0.889528     0.936402  0.912364  
8         0.856988     0.716838  0.780673  
9         0.885208     0.922471  0.903455  
10        0.854197     0.702604  0.771020  
11        0.904950     0.276802  0.423933

**Analysis**:
- RandomForestClassifier + TfidfVectorizer best results are with unigrams being these a bit bellow the other model (Accuracy: 0.932, Recall: 0.936)

### 5. Current Best ML Model Validation

After seeing the results of the premade models and given the best results we got were based on a LogisticRegression we assume there are some words in the dataset hevily influencing the model.

In [20]:
# Let's first train the logistic regressor
lr = sklearn.linear_model.LogisticRegression(random_state=seed)
cv = sklearn.feature_extraction.text.CountVectorizer(ngram_range = (1, 2))

mv_train_df = cv.fit_transform(X_train)
mv_test_df  = cv.transform(X_test)

lr.fit(mv_train_df, y_train)

y_train_predict = lr.predict(mv_train_df)
y_test_predict = lr.predict(mv_test_df)

result = get_classifier_metrics(y_train, y_train_predict , sub_name="_train") | get_classifier_metrics(y_test, y_test_predict , sub_name="_test") 
display(pd.DataFrame.from_dict(result, orient='index').T)

# Get the coeficients and the features from the model and vectorizer
coefs = lr.coef_[0]
features = cv.get_feature_names_out()

accuracy_train  precision_train  recall_train  f1_train  accuracy_test  \
0        0.998243         0.996845      0.999548  0.998195       0.952862   

   precision_test  recall_test   f1_test  
0        0.940828     0.963053  0.951811

In [21]:
# Review the most influencial words for the predictions
word_coefs = pd.DataFrame(zip(features, coefs))
word_coefs.columns = ["word", "coef"]

# See the top 10 words with more coeficient
display(word_coefs.sort_values("coef", ascending=False)[0:10])

# See the top 10 words with less coeficient
display(word_coefs.sort_values("coef", ascending=True)[0:10])

word      coef
49321     factbox  3.182219
127033       says  2.709662
157549      urges  2.001461
147551  tillerson  1.741403
82476   lawmakers  1.733302
141728      talks  1.730247
26569       china  1.727194
47074          eu  1.723963
168852   zimbabwe  1.693987
136309  spokesman  1.622523

word      coef
158962      video -4.307516
19445    breaking -4.102125
60141         gop -3.593094
66407     hillary -3.400183
79005        just -3.086532
117247     racist -2.433639
38121         dem -2.108554
69683        huge -2.043482
167531        wow -1.976108
18423   bombshell -1.911474

In [22]:
# Let's try the model by removing parts of it
def remove_words(serie, words):
    return serie.apply(lambda x: ' '.join([w for w in x.split() if w not in words]))

results = []
for i in range(5, 41, 5):
    c = i / 10
    words = list(word_coefs.loc[(word_coefs["coef"] > c) | (word_coefs["coef"] < c * -1), "word"])
    y_test_predict = lr.predict(cv.transform(remove_words(X_test, words)))
    result["coef_thld_rm"] = c 
    result["num_words_rm"] = len(words)
    result["prop_words_rm"] = len(words) / word_coefs.shape[0]
    results.append(result | get_classifier_metrics(y_test, y_test_predict , sub_name="_test"))

results = pd.DataFrame(results).sort_values("coef_thld_rm", ascending=False)
display(results)

accuracy_train  precision_train  recall_train  f1_train  accuracy_test  \
7        0.998243         0.996845      0.999548  0.998195       0.940419   
6        0.998243         0.996845      0.999548  0.998195       0.937052   
5        0.998243         0.996845      0.999548  0.998195       0.927536   
4        0.998243         0.996845      0.999548  0.998195       0.925194   
3        0.998243         0.996845      0.999548  0.998195       0.924608   
2        0.998243         0.996845      0.999548  0.998195       0.907627   
1        0.998243         0.996845      0.999548  0.998195       0.844971   
0        0.998243         0.996845      0.999548  0.998195       0.701069   

   precision_test  recall_test   f1_test  coef_thld_rm  num_words_rm  \
7        0.917027     0.963961  0.939908           4.0             2   
6        0.910990     0.963961  0.936727           3.5             3   
5        0.894574     0.963658  0.927832           3.0             6   
4        0.898828     0.952453  0.924864           2.5             7   
3        0.897575     0.952756  0.924343           2.0            11   
2        0.871282     0.949122  0.908537           1.5            43   
1        0.784140     0.937311  0.853911           1.0           222   
0        0.625498     0.950939  0.754626           0.5          1371   

   prop_words_rm  
7       0.000012  
6       0.000018  
5       0.000035  
4       0.000041  
3       0.000065  
2       0.000254  
1       0.001313  
0       0.008111

**Analysis**:
- After removing the words iterativelly based on the coeficient of the words we see a decrease in accuracy (0.844 by keeping only words with coeficient inbetween -1 and 1)
- The recall stays relevant in between 0.937 and 0.963 
- This test shows that the model is 

## 6. Pre-Trained Transformers models experimentation

### 6.1. Bert tiny fine-tunned fake news detector

In [23]:

# FAKE NEWS DETECTOR
pipe = pipeline("text-classification", model="mrm8488/bert-tiny-finetuned-fake-news-detection", device=0)

predictions = []
for text in train_df["corpus"]:
    predictions.append(0 if pipe(text)[0]["label"] == "LABEL_0" else 1)

results = get_classifier_metrics(train_df["label"], predictions)
results = {"model": "mrm8488/bert-tiny-finetuned-fake-news-detection"} | results
results = pd.DataFrame.from_dict(results, orient='index').T
display(results)

Device set to use cuda:0
You seem to be using the pipelines sequentially on GPU. In order to maximize efficiency please use a dataset


model  accuracy precision  \
0  mrm8488/bert-tiny-finetuned-fake-news-detection  0.465214  0.474819   

   recall        f1  
0  0.9576  0.634851

### 6.2. Fake news Bert detect

In [24]:
MODEL_bert = "jy46604790/Fake-News-Bert-Detect"
clf_bert = pipeline("text-classification", model=MODEL_bert, tokenizer=MODEL_bert, device=0)

text_bert = train_df["corpus"].tolist()
result_bert = clf_bert(text_bert)
predictions_bert = [0 if r["label"] == "LABEL_0" else 1 for r in result_bert]

result = get_classifier_metrics(train_df["label"], predictions_bert)
result = {"model": MODEL_bert} | result
result = pd.DataFrame.from_dict(result, orient='index').T

display(result)
results = pd.concat([results, result], ignore_index=True)

Device set to use cuda:0


model  accuracy precision    recall        f1
0  jy46604790/Fake-News-Bert-Detect  0.652319  0.983162  0.288782  0.446434

### 6.3. Distil Bert fake news

In [25]:
MODEL_dist = "yasmine-11/distilbert_fake_news"
clf_dist = pipeline("text-classification", model=MODEL_dist, tokenizer=MODEL_dist, device=0)

text_dist = X_train.tolist()
result_dist = clf_dist(text_dist)
predictions_dist = [0 if r["label"] == "LABEL_0" else 1 for r in result_dist]

result = get_classifier_metrics(y_train, predictions_dist)
result = {"model": MODEL_dist} | result
result = pd.DataFrame.from_dict(result, orient='index').T

display(result)
results = pd.concat([results, result], ignore_index=True)

Device set to use cuda:0


model  accuracy precision    recall        f1
0  yasmine-11/distilbert_fake_news  0.484426  0.485065  0.988176  0.650714

In [26]:
results.to_csv("./output/pre_trained_models.csv")
display(results)

model  accuracy precision  \
0  mrm8488/bert-tiny-finetuned-fake-news-detection  0.465214  0.474819   
1                 jy46604790/Fake-News-Bert-Detect  0.652319  0.983162   
2                  yasmine-11/distilbert_fake_news  0.484426  0.485065   

     recall        f1  
0    0.9576  0.634851  
1  0.288782  0.446434  
2  0.988176  0.650714

### 7. Data Cleaning Alternative Strategies

In [27]:
def less_cleaning(serie):
    # Replace apostrofo
    serie = serie.replace(r",s", "'s", regex=True, inplace=False)
    # Remove SOME characters, keeping * ' - ( ) 
    serie = serie.replace(r"[^a-zA-Z0-9*'\-().]", " ", regex=True, inplace=False)
    # Remove multiple empty spaces
    serie = serie.replace(r"\s{2,}", " ", regex=True, inplace=False)
    # Make all string lowercase
    serie = serie.str.lower()
    return serie

X2_train, X2_test, y2_train, y2_test = sklearn.model_selection.train_test_split(train_df["corpus"], train_df["label"], test_size=.2, random_state=seed)

results=[]
result = evaluate_model(model = sklearn.linear_model.LogisticRegression(random_state=seed), 
                        vectorizer = sklearn.feature_extraction.text.CountVectorizer(ngram_range = (1, 2)),
                        X_train    = less_cleaning(X2_train), 
                        y_train    = y2_train, 
                        X_test     = less_cleaning(X2_test), 
                        y_test     = y2_test)
result["name"] = "less_cleaning"
results.append(result)

result = evaluate_model(model      = sklearn.linear_model.LogisticRegression(random_state=seed), 
                        vectorizer = sklearn.feature_extraction.text.CountVectorizer(ngram_range = (1, 2)),
                        X_train    = clean_serie(X2_train), 
                        y_train    = y2_train, 
                        X_test     = clean_serie(X2_test), 
                        y_test     = y2_test)
result["name"] = "clean_serie"
results.append(result)

result = evaluate_model(model      = sklearn.linear_model.LogisticRegression(random_state=seed), 
                        vectorizer = sklearn.feature_extraction.text.CountVectorizer(ngram_range = (1, 2)),
                        X_train    = X2_train, 
                        y_train    = y2_train, 
                        X_test     = X2_test, 
                        y_test     = y2_test)
result["name"] = "no_cleanning"
results.append(result)

results = pd.DataFrame(results)
results.to_csv("./output/alternative_data_cleanning.csv")
display(results)

model  fit_time       vectorizer  accuracy_train  \
0  LogisticRegression  5.339275  CountVectorizer        0.998426   
1  LogisticRegression  4.540784  CountVectorizer        0.998243   
2  LogisticRegression  5.126765  CountVectorizer        0.998426   

   precision_train  recall_train  f1_train  accuracy_test  precision_test  \
0         0.997220      0.999548  0.998383       0.952716        0.940290   
1         0.996845      0.999548  0.998195       0.952862        0.940828   
2         0.997220      0.999548  0.998383       0.952569        0.939752   

   recall_test   f1_test           name  
0     0.963356  0.951683  less_cleaning  
1     0.963053  0.951811    clean_serie  
2     0.963658  0.951555   no_cleanning

## 8. Final Model Application

### 8.1. Best Model Configuration

In [28]:
# Best Model
bmodel = sklearn.linear_model.LogisticRegression(random_state=seed)
bvect  = sklearn.feature_extraction.text.CountVectorizer(ngram_range = (1, 2))

# Train the model with all data
bm_train_ds = bvect.fit_transform(clean_serie(train_df["corpus"]))

# Fit the model
bmodel.fit(bm_train_ds, train_df["label"])

bmodel_train_predict = bmodel.predict(bm_train_ds)

result = get_classifier_metrics(train_df["label"], bmodel_train_predict , sub_name="_train")
display(pd.DataFrame.from_dict(result, orient='index').T)

accuracy_train  precision_train  recall_train  f1_train
0        0.998536         0.997352      0.999638  0.998494

In [29]:
val_df = pd.read_csv("./dataset/testing_data.csv", sep="\t", header=None)
val_df.columns = ["label", "corpus"]

X_val = clean_serie(val_df["corpus"])
val_ds = bvect.transform(X_val)

val_predict = bmodel.predict(val_ds)

### 8.2. Save the results in a new file

In [30]:
val_df["label"] = pd.Series(val_predict)
display(val_df)
val_df.to_csv("./output/testing_data.csv", sep="\t", header=False, index=False)

label                                             corpus
0         0  copycat muslim terrorist arrested with assault...
1         0  wow! chicago protester caught on camera admits...
2         1   germany's fdp look to fill schaeuble's big shoes
3         0  mi school sends welcome back packet warning ki...
4         1  u.n. seeks 'massive' aid boost amid rohingya '...
...     ...                                                ...
9979      0  boom! fox news leftist chris wallace attempts ...
9980      0  here it is: list of democrat hypocrites who vo...
9981      1  new fires ravage rohingya villages in northwes...
9982      0  meals on wheels shuts the lyin‚ lefties up wit...
9983      0  brilliant! tucker carlson and ayaan hirsi ali ...

[9984 rows x 2 columns]